swift/StableDiffusion/tokenizer/T5Tokenizer.swift

```diff
@@ -0,0 +1,1975 @@
+        // https://github.com/xenova/transformers.js/blob/main/src/tokenizers.js#L2237-L2247
+        // - For a proper implementation, see:
+        // https://github.com/huggingface/tokenizers/blob/b58227c7f1ccf8b73ee2268354336da56d91e492/tokenizers/src/normalizers/precompiled.rs#L36
+        var output: String = ""
+        var hasFullwidthTilde = false
+
+        for scalar in text.unicodeScalars {
+            switch scalar.value {
+            case 0x0001...0x0008, 0x000B, 0x000E...0x001F, 0x007F, 0x008F, 0x009F:
+                // Non-printing control characters
+                output.append("")
+            case 0x0009, 0x000A, 0x000C, 0x000D, 0x1680, 0x200B...0x200F, 0x2028, 0x2029, 0x2581, 0xFEFF, 0xFFFD:
+                // Separators
+                output.append(" ")
+            case 0xFF5E:
+                hasFullwidthTilde = true
+                fallthrough
+            default:
+                output.append(Character(scalar))
+            }
+        }
+
+        if hasFullwidthTilde {
+            return output
+                .split(by: "\u{FF5E}")
+                .map({ $0.precomposedStringWithCompatibilityMapping })
+                .joined(separator: "\u{FF5E}")
+        } else {
+            return output.precomposedStringWithCompatibilityMapping
+        }
+    }
+}
+
+class StripAccentsNormalizer: Normalizer {
+    required init(config: Config) {}
+
+    func normalize(text: String) -> String {
+        text.precomposedStringWithCompatibilityMapping
+    }
+}
+
+enum StringReplacePattern {
+    case regexp(regexp: NSRegularExpression, replacement: String)
+    case string(pattern: String, replacement: String)
+}
+
+extension StringReplacePattern {
+    func replace(_ text: String) -> String {
+        switch self {
+        case .regexp(let regexp, let replacement):
+            let range = NSRange(text.startIndex..., in: text)
+            let replaced = regexp.stringByReplacingMatches(in: text, options: [], range: range, withTemplate: replacement)
+            return replaced
+        case .string(let toReplace, let replacement):
+            return text.replacingOccurrences(of: toReplace, with: replacement)
+        }
+    }
+}
+
+extension StringReplacePattern {
+    static func from(config: Config) -> StringReplacePattern? {
+        guard let replacement = config.content?.stringValue else { return nil }
+        if let pattern = config.pattern?.String?.stringValue {
+            return StringReplacePattern.string(pattern: pattern, replacement: replacement)
+        }
+        if let pattern = config.pattern?.Regex?.stringValue {
+            guard let regexp = try? NSRegularExpression(pattern: pattern, options: []) else {
+                fatalError("Cannot build regexp from \(pattern)")
+            }
+            return StringReplacePattern.regexp(regexp: regexp, replacement: replacement)
+        }
+        return nil
+    }
+}
+
+let byteEncoder: Dictionary<UTF8.CodeUnit, String> = [
```

Comment:
Thanks for the review @alejandro-isaza, as a general note, this T5Tokenizer code is actually copied/adapted from [swift-transformers](https://github.com/argmaxinc/ml-stable-diffusion/blob/d1f0604fab5345011e0b9f5b87ee0c155612565f/swift/StableDiffusion/tokenizer/T5Tokenizer.swift#L1). We may want to simply add swift-transformers as a dependency, and we can bring these suggestions over there. Curious your thoughts, otherwise I will be happy to adapt this code further based on your notes.

Reply:
I think adding a dependency is a good idea. Ideally it would be swift-tokenizers :)